### PR TITLE
add support members with expiration for role and group

### PR DESCRIPTION
**Pull request by mendi160**
Signed-off-by: mshneorson <mendi.shneorson@yahooinc.com>

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -43,4 +43,6 @@ docker/sample/domain-admin/domain_admin_cert.pem
 docker/sample/domain-admin/domain_admin_key.pem
 docker/zms/conf/zms.properties
 
-manual-testing/
\ No newline at end of file
+manual-testing/
+
+/example
```
Comment (dvirguttman) [inline .gitignore:48]: pls avoid adding unnecessary folders here unless it is really needed. 
and if it is, consider change it to local-testing or something similar

File: athenz/consts.go [hunk 2]
```diff
@@ -9,4 +9,6 @@ const (
 	SERVICE_SEPARATOR    = "."
 	SUB_DOMAIN_SEPARATOR = "."
 	PREFIX_USER_DOMAIN   = "home."
+	EXPIRATION_TEMPLATE  = "2006-01-02 15:04:05" // General type
```
Comment (dvirguttman) [inline athenz/consts.go:12]: What do you mean general type?
why is this specific date?

File: docs/resources/athenz_role.md [hunk 3]
```diff
@@ -35,7 +59,13 @@ The following arguments are supported:
...
     
 
-- `members` - (Optional) List of Athenz principal members. must be in this format: `user.<userid> or <domain>.<service> or <domain>:group.<group>`.
+- `members` - \*\*Deprecated** use member attribute instead (Optional) List of Athenz principal members. must be in this format: `user.<userid> or <domain>.<service> or <domain>:group.<group>`.
+
+- `member` - (Optional) A set of Athenz principal members. Each member consists the following arguments:
+
+  - `name` - (Required) The name of the Athenz principal member. must be in this format: `user.<userid> or <domain>.<service> or <domain>:group.<group>`.
+
+  - `expiration` - (Optional) The expiration of the Athenz principal member. must be in this format: `<yyyy>-<mm>-<dd> <hh>:<MM>:<ss>`
```
Comment (mendi160) [inline docs/resources/athenz_role.md:68]: @dvirguttman Do you think the seconds are necessary?

File: athenz/resource_role.go [hunk 4]
```diff
@@ -2,12 +2,12 @@ package athenz
...
 import (
 	"context"
-	"log"
-
 	"github.com/AthenZ/athenz/clients/go/zms"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"log"
 
 	"github.com/AthenZ/terraform-provider-athenz/client"
+	//_ "github.com/hashicorp/terraform-plugin-framework/resource"
```
Comment (dvirguttman) [inline athenz/resource_role.go:10]: make sure to run go fmt and goimports

File: athenz/resource_role.go [hunk 5]
```diff
@@ -36,11 +36,33 @@ func ResourceRole() *schema.Resource {
...
-				Description: "Users or services to be added as members",
-				Optional:    true,
-				Elem:        &schema.Schema{Type: schema.TypeString},
-				Set:         schema.HashString,
+				Type:          schema.TypeSet,
+				Description:   "Users or services to be added as members",
+				Optional:      true,
+				Elem:          &schema.Schema{Type: schema.TypeString},
+				Set:           schema.HashString,
+				ConflictsWith: []string{"member"},
```
Comment (dvirguttman) [inline athenz/resource_role.go:44]: is it possible to create test for conflict case?

File: athenz/resource_role.go [hunk 6]
```diff
@@ -169,24 +201,76 @@ func resourceRoleRead(ctx context.Context, d *schema.ResourceData, meta interfac
 	return nil
 }
 
+// return true iff the members is includes in the members list
+func isMembersIncludes(member *zms.RoleMember, members []*zms.RoleMember) bool {
```
Comment (dvirguttman) [inline athenz/resource_role.go:205]: `isMember`

File: athenz/resource_role.go [hunk 7]
```diff
@@ -169,24 +201,76 @@ func resourceRoleRead(ctx context.Context, d *schema.ResourceData, meta interfac
...
+	for _, m := range members {
+		if string(member.MemberName) == string(m.MemberName) {
+			return true
+		}
+	}
+	return false
+}
+
+// return all role members that appears in list1 but not includes in list2
+func unifyMembers(list1, list2 []*zms.RoleMember) []*zms.RoleMember {
```
Comment (dvirguttman) [inline athenz/resource_role.go:215]: is is not unify, but something like `retainAll` in java

File: athenz/resource_role.go [hunk 8]
```diff
@@ -169,24 +201,76 @@ func resourceRoleRead(ctx context.Context, d *schema.ResourceData, meta interfac
...
+	if len(removeDeprecatedRoleMembers) == 0 {
+		removeMembers = unifyMembers(removeRoleMembers, addDeprecatedRoleMembers)
+	} else {
+		removeMembers = unifyMembers(removeDeprecatedRoleMembers, addRoleMembers)
+	}
+	if len(addDeprecatedRoleMembers) == 0 {
+		return removeMembers, addRoleMembers
+	}
+	return removeMembers, addDeprecatedRoleMembers
+}
```
Comment (dvirguttman) [inline athenz/resource_role.go:241]: not sure what we are trying to do here

File: athenz/resource_role.go [hunk 9]
```diff
@@ -169,24 +201,76 @@ func resourceRoleRead(ctx context.Context, d *schema.ResourceData, meta interfac
...
+	if d.HasChange("member") {
+		if _, ok := d.GetOk("trust"); ok {
+			return diag.Errorf("delegated roles cannot change members")
 		}
+		os, ns := handleChange(d, "member")
+		removeRoleMembers = expandRoleMembers(os.Difference(ns).List())
+		addRoleMembers = expandRoleMembers(ns.Difference(os).List())
+	}
+	removeMembers, addMembers := handleMembersChange(removeDeprecatedRoleMembers, addDeprecatedRoleMembers, removeRoleMembers, addRoleMembers)
+	err = updateRoleMembers(dn, rn, removeMembers, addMembers, auditRef, zmsClient)
```
Comment (dvirguttman) [inline athenz/resource_role.go:271]: It is too complicated (and might lead to unexpected behavior).
I would suggest to split it to 2 different methods - `membersToAdd` membersToDelete`

File: athenz/resource_role.go (hunk 8 above)
Comment (mendi160) [reply athenz/resource_role.go:241]: I want to make sure not to delete unnecessary members when the users switch between members attribute to member attribute and vice versa

File: athenz/utils_group.go [hunk 10]
```diff
@@ -5,10 +5,48 @@ import (
...
+	} else {
+		removeMembers = unifyGroupMembers(removeDeprecatedRoleMembers, addRoleMembers)
+	}
+	if len(addDeprecatedRoleMembers) == 0 {
+		return removeMembers, addRoleMembers
+	}
+	return removeMembers, addDeprecatedRoleMembers
+}
+
+func expandDeprecatedGroupMembers(configured []interface{}) []*zms.GroupMember {
```
[the remark this answers is not in the document]
Comment (mendi160) [reply athenz/utils_group.go:49]: no. it's not just additional attribute, it's a different set type (  set of strings vs set of set of map[string]interface{})

File: athenz/resource_role.go [hunk 11]
```diff
@@ -169,24 +201,76 @@ func resourceRoleRead(ctx context.Context, d *schema.ResourceData, meta interfac
 	return nil
 }
 
+// return true iff the members is includes in the members list
```
[the remark this answers is not in the document]
Comment (mendi160) [reply athenz/resource_role.go:204]: if and only if

File: athenz/consts.go [hunk 12]
```diff
@@ -9,4 +9,6 @@ const (
 	SERVICE_SEPARATOR    = "."
 	SUB_DOMAIN_SEPARATOR = "."
 	PREFIX_USER_DOMAIN   = "home."
+	EXPIRATION_LAYOUT    = "2006-01-02 15:04:05"
```
[the remark this answers is not in the document]
Comment (mendi160) [reply athenz/consts.go:12]: This is the earliest possible date that time.ParseInLocation can get as layout

File: athenz/data_source_group.go [hunk 13]
```diff
@@ -53,7 +65,9 @@ func dataSourceGroupRead(ctx context.Context, d *schema.ResourceData, meta inter
 	d.SetId(fullResourceName)
 
 	if len(group.GroupMembers) > 0 {
-		d.Set("members", flattenGroupMember(group.GroupMembers))
+		if err = d.Set("member", flattenGroupMembers(group.GroupMembers)); err != nil {
+			return diag.FromErr(err)
+		}
```
[the remark this answers is not in the document]
Comment (mendi160) [reply athenz/data_source_group.go:70]: data source...

File: athenz/data_source_group.go [hunk 14]
```diff
@@ -22,11 +22,23 @@ func DataSourceGroup() *schema.Resource {
 				Type:     schema.TypeString,
 				Required: true,
 			},
-			"members": {
-				Type:     schema.TypeSet,
-				Optional: true,
-				Elem:     &schema.Schema{Type: schema.TypeString},
-				Set:      schema.HashString,
+			"member": {
```
[the remark this answers is not in the document]
Comment (mendi160) [reply athenz/data_source_group.go:25]: this is data_source...

File: athenz/data_source_group.go [hunk 15]
```diff
@@ -22,11 +22,23 @@ func DataSourceGroup() *schema.Resource {
...
+				Type:        schema.TypeSet,
+				Description: "Users or services to be added as members",
```
[the remark this answers is not in the document]
Comment (mendi160) [reply athenz/data_source_group.go:27]: A group can't be a member of another group. I will change the description for role

File: athenz/consts.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (dvirguttman) [reply athenz/consts.go:12]: so maybe choose some generic date, it looks very specific

File: athenz/resource_role.go [hunk 16]
```diff
@@ -169,25 +201,84 @@ func resourceRoleRead(ctx context.Context, d *schema.ResourceData, meta interfac
...
+		}
+	}
+	return toReturn
+}
+
+/*
+since we not allow configuring both members (deprecated) and member attributes at the same time, state changes CAN'T be one of the following:
+1. remove members from both attributes member and members.
+2. add members in both attributes member and members.
+*/
```
Comment (dvirguttman) [inline athenz/resource_role.go:229]: I am not sure it's correct.
what about this use case for example:
1. the user has old `members` block defined with `member-1,member-2`.
2. the user is going to change the TF file to `member` block with `member-1` only

Comment (mendi160) [reply athenz/resource_role.go:229]: so in this case the change in the state for members attributes is to delete member 1 and member 2  and the changes In the state for member attribute is to add member 2.

Comment (mendi160) [reply athenz/resource_role.go:229]: this what will happen: we will perform a delete api call for member1, and will perform add api call for member2 (will cause override)

**Review by dvirguttman — APPROVED** [review]

**Pull request merged**